pkg/runner/resume.go

```diff
@@ -0,0 +1,270 @@
+						}
+					}
+				}
+			}
+		}
+
+		return nil
+	}
+
+	goroutineManager := manager.NewGoroutineManager(
+		ctx,
+		&errs,
+		manager.GoroutineManagerHooks{
+			OnAfterRecover: func() {
+				if suspendOnPanicWithError {
+					suspendCtx, cancelSuspendCtx := context.WithTimeout(runner.rescueCtx, rescueTimeout)
+					defer cancelSuspendCtx()
+
+					// Connections need to be closed before creating the snapshot
+					if resumedRunner.acceptingAgent != nil {
+						if e := resumedRunner.acceptingAgent.Close(); e != nil {
+							errs = errors.Join(errs, snapshotter.ErrCouldNotCloseAcceptingAgent, e)
+						}
+					}
+					if resumedRunner.agent != nil {
+						resumedRunner.agent.Close()
+					}
+
+					// If a resume failed, flush the snapshot so that we can re-try
+					if err := resumedRunner.createSnapshot(suspendCtx); err != nil {
+						errs = errors.Join(errs, ErrCouldNotCreateRecoverySnapshot, err)
+					}
+				}
+			},
+		},
+	)
+	defer goroutineManager.Wait()
+	defer goroutineManager.StopAllGoroutines()
+	defer goroutineManager.CreateBackgroundPanicCollector()()
+
+	// We intentionally don't call `wg.Add` and `wg.Done` here since we return the process's wait method
+	// We still need to `defer handleGoroutinePanic()()` here however so that we catch any errors during this call
+	goroutineManager.StartBackgroundGoroutine(func(_ context.Context) {
+		if err := runner.server.Wait(); err != nil {
+			panic(errors.Join(ErrCouldNotWaitForFirecracker, err))
+		}
+	})
+
+	var err error
+	resumedRunner.agent, err = ipc.StartAgentServer(
+		filepath.Join(runner.server.VMPath, snapshotter.VSockName),
+		uint32(agentVSockPort),
+	)
+	if err != nil {
+		panic(errors.Join(snapshotter.ErrCouldNotStartAgentServer, err))
+	}
+
+	resumedRunner.Close = func() error {
+		resumedRunner.agent.Close()
+
+		return nil
+	}
+
+	if err := os.Chown(resumedRunner.agent.VSockPath, runner.hypervisorConfiguration.UID, runner.hypervisorConfiguration.GID); err != nil {
+		panic(errors.Join(ErrCouldNotChownVSockPath, err))
+	}
+
+	{
+		resumeSnapshotAndAcceptCtx, cancelResumeSnapshotAndAcceptCtx := context.WithTimeout(goroutineManager.Context(), resumeTimeout)
+		defer cancelResumeSnapshotAndAcceptCtx()
+
+		if err := firecracker.ResumeSnapshot(
+			resumeSnapshotAndAcceptCtx,
+
+			runner.firecrackerClient,
+
+			runner.stateName,
+			runner.memoryName,
+
+			!snapshotLoadConfiguration.ExperimentalMapPrivate,
+		); err != nil {
+			panic(errors.Join(ErrCouldNotResumeSnapshot, err))
+		}
+
+		suspendOnPanicWithError = true
+
+		resumedRunner.acceptingAgent, err = resumedRunner.agent.Accept(resumeSnapshotAndAcceptCtx, ctx)
+		if err != nil {
+			panic(errors.Join(ErrCouldNotAcceptAgent, err))
+		}
+	}
+
+	// We intentionally don't call `wg.Add` and `wg.Done` here since we return the process's wait method
+	// We still need to `defer handleGoroutinePanic()()` here however so that we catch any errors during this call
```

Comment:
```suggestion
```

Reply:
Resolved in 88ddb283155e97e3fdae49179ed5c9e3a1c63b23